lib/block_view.go

```diff
@@ -2936,24 +3023,95 @@ func _isEntryImmatureBlockReward(utxoEntry *UtxoEntry, blockHeight uint32, param
 	if err != nil {
-		return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		return errors.Wrapf(err, "_verifySignature: Problem parsing signature: ")
 	}
-	// Verify that the transaction is signed by the specified key.
-	if txn.Signature == nil || !txn.Signature.Verify(txHash[:], txnPk) {
-		return RuleErrorInvalidTransactionSignature
+
+	// We check for public key solution iter for this transaction. This
+	// allows us to recover public key from the signature.
+	// If the transaction was signed with the owner key, iter is equal to 0.
+	// If the transaction was signed with a derived key, iter is between 1-4.
+	iter := sigByte - DERControlByte
+	if iter == 0 {
+		// Convert the txn public key into a *btcec.PublicKey
+		txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+		if err != nil {
+			return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		}
+		// Verify that the transaction is signed by the specified key.
+		if sig.Verify(txHash[:], txnPk) {
+			return nil
+		}
+	} else {
+		// Subtract 1 from iter because iteration should be between 0-3.
+		iter = iter - 1
+
+		// Recover public key from the signature
+		sigCompact := SignatureSerializeCompactWithIter(sig, int(iter), false)
+		derivedPublicKey, _, err := btcec.RecoverCompact(btcec.S256(), sigCompact, txHash[:])
+		if err != nil {
+			return errors.Wrapf(err, "_verifySignature: Problem recovering public key ")
+		}
+
+		// Get owner PKID
+		ownerPKID := PublicKeyToPKID(txn.PublicKey)
+		// Get derived PKID
+		derivedPKID := PublicKeyToPKID(derivedPublicKey.SerializeCompressed())
+
+		// We check if there's an entry in UtxoView that authorizes this derived key
+		if entryMap, ok := bav.PKIDToDerivedKeyEntry[*ownerPKID]; ok {
+			if entry, ok := entryMap[*derivedPKID]; ok {
+				// Check if the key hasn't been revoked
+				if entry != nil {
+					if entry.OperationType == AuthorizeDerivedKeyOperationValid && entry.ExpirationBlock > uint64(blockHeight) {
+						return nil
+						// I'm almost certain this is redundant since we've already recovered the public key.
+						//if sig.Verify(txHash[:], derivedPublicKey) {
+						//	return nil
+						//}
+					} else if entry.OperationType == AuthorizeDerivedKeyOperationInValid {
+						// If the owner de-authorized a key in this block, we want to immediately prevent any
+						// transactions signed with the derived key. So we return and skip Db check.
+						return RuleErrorInvalidTransactionSignature
+					}
+				}
+			}
+		}
+
+		// We check if the derived key exists in DB and hasn't expired.
+		authEntry := DBGetOwnerToDerivedKeyMapping(bav.Handle, ownerPKID, derivedPKID)
+		if authEntry.OperationType == AuthorizeDerivedKeyOperationValid && authEntry.ExpirationBlock > uint64(blockHeight){
+			return nil
+			// I'm almost certain this is redundant since we already recovered the public key.
+			//if sig.Verify(txHash[:], derivedPublicKey) {
+			//	return nil
+			//}
+		}
```

Comment:
This should be a helper on `UtxoView` and the result from the database should be cached in the `PKIDToDerivedKeyEntry` map